<!DOCTYPE html PUBLIC "-//IETF//DTD HTML//EN">
<!-- saved from url=(0063)http://web.expasy.org/peptide_cutter/peptidecutter_enzymes.html -->
<html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">

    <!-- automatically added by Apache -->
    <link rel="shortcut icon" href="http://web.expasy.org/favicon.ico"><!-- automatically added by Apache -->
    <link rel="stylesheet" href="./expasy_peptidecuttr_enzymes_files/sib.css" type="text/css" media="screen" charset="utf-8">
	   <link rel="stylesheet" href="./expasy_peptidecuttr_enzymes_files/sib_print.css" type="text/css" media="print" charset="utf-8">
           <!--[if lte IE 6]>
	   <link rel="stylesheet" href="/css/sib_css/sib_ie6.css" type="text/css" media="screen" title="" charset="utf-8">
           <![endif]-->
           <link rel="stylesheet" href="./expasy_peptidecuttr_enzymes_files/base.css" type="text/css"><!-- automatically added by Apache -->
    <title>ExPASy PeptideCutter tool: available enzymes</title><!-- 1swiss-prot 1tools -->

  </head>



<body bgcolor="#FFFFFF" text="#000000">

<!-- ********************************************************************************************** -->
<!-- START OF HEADER information on top of the page -->
<div id="sib_top"><a name="TOP"></a></div>
<div id="sib_container">

<!-- TODO: Change id to "sib_header" or "sib_header_small" -->
<div id="sib_header_small">

<!-- SIB-ExPASy -->
<div id="sib_expasy_logo">
<a href="http://www.isb-sib.ch/" id="sib_link" title="SIB Swiss Institute of Bioinformatics"></a>
<a href="http://www.expasy.org/" id="expasy_link" title="ExPASy - SIB Bioinformatics Resource Portal"></a>
</div>
<!-- SIB-ExPASy END -->

<div class="sib_title_no_logo">
<h1><a href="http://web.expasy.org/peptide_cutter">PeptideCutter</a></h1>
</div>

<!-- -- Additionnal resource header ---->
<div id="resource_header"></div>

<div id="sib_header_nav">
<a href="http://web.expasy.org/peptide_cutter" title="Home page">Home</a>&nbsp;|&nbsp;<a style="font-weight: bold; color: rgb(204,0,0) !important;" href="http://web.expasy.org/contact" title="Contact">Contact</a></div>
</div>
<!-- END OF HEADER -->
<!-- ********************************************************************************************** -->
<div id="sib_body">
<!-- automatically added by Apache -->
<hr size="2">



<h1>

 <a href="http://web.expasy.org/peptide_cutter/">PeptideCutter</a><p>

</p></h1>



    <h1> The cleavage specificities of selected enzymes and chemicals:</h1><hr>



<h2> A general model of enzymatic cleavage:</h2><br><br>

<center><img align="MIDDLE" src="./expasy_peptidecuttr_enzymes_files/protein_cleavage.jpg" height="288" width="576" t="ExPASy Logo"></center>

<br><br>

Subsite nomenclature was adopted from a scheme created by <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Schechter1967">Schechter 

and Berger (1967, 1968)</a> and used in the following description of enzyme specificities. 

According to this model, amino acid residues in a substrate undergoing cleavage 

are designated P1,P2, P3, P4 etc. in the N-terminal direction from the cleaved 

bond. Likewise, the residues in C-terminal direction are designated P1', P2', 

P3', P4'. etc. as shown in <b>Fig.1</b>. Most of the following rules of protease 

specificity were adopted from the comprehensive publication of <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">Keil 

( 1992)</a>. These rules were verified by checking against information published 

by<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Barrett"> Barrett et al. (1998)</a>. 

<br>

<br><br>



<hr><br><br>





<h2> Establishing a specificity model for protease cleavage:</h2><br><hr><br>

A first and obvious approach to obtain information concerning the cleavage specificity 

of a protease is the characterization of the respective natural substrate. In 

a next step, standard polypeptides can be used for digestion such as glucagon 

or insulin chains. However, an ideal polypeptidic substrate should contain all 

possible 400 combinations of dipeptide bonds, whereas in insulin and glucagon 

for example only small fractions of these combinations can be found (not to mention 

the possibilities of tetrapeptide composition when taking into account the sites 

P2 to P2'). Ideally, the cleavage data would be obtained by digesting all availble 

proteins in the databases, but this is beyond feasability. A more systematic and 

complete approach would be to test the proteases with substrates of low molecular 

weight such as di-, tri, tetrapeptides etc.<br>

<br>

Unfortunately, the available data for most proteases is still very incomplete. 

Only for<a href="http://web.expasy.org/peptide_cutter/peptidecutter_special_enzymes.html"> a few proteases </a>enough 

information has been accumulated that allow a statistical treatment (for details 

see <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">Keil (1992)</a> resulting 

in a more complete and refined picture of cleavage specificity. In the following, 

specific cleavage preferences of individual enzymes are reported. Only the accordingly 

derived rules are taken into consideration by the program PeptideCutter. Thus, 

the user should be aware of the fact, that results obtained by experiments may 

differ from the predictions made by the PeptideCutter program. <br>

<br>

<hr>

<h2> The cleavage specificities of selected enzymes and chemicals:</h2>

<br>

<hr>

<h2><a name="ArgC"></a><a href="http://enzyme.expasy.org/EC/3.4.21.35">Arg-C proteinase:</a></h2>

The Arg-C proteinase preferentially cleaves at Arg in position P1. The cleavage 

behaviour seems to be only moderately affected by residues in position P1' (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">Keil, 

1992</a>). <br>

<hr>







<h2><a name="AspN"></a><a href="http://enzyme.expasy.org/EC/3.4.24.33">Asp-N Endopeptidase:</a></h2>

The Asp-N Endopeptidase cleaves specifically bonds with Asp in position P1' (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">Keil, 1992</a>).

<br><hr>



<h2><a name="AspGluN"></a><a href="http://enzyme.expasy.org/EC/3.4.24.33">Asp-N Endopeptidase + N-terminal Glu:</a></h2>

The Asp-N Endopeptidase cleaves specifically bonds with Asp or Glu in position P1' (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">Keil, 1992</a>).

<br><hr>



<h2><a name="BNPS"></a>BNPS-Skatole:</h2>

BNPS-skatole [2-(2-nitrophenylsulfenyl)-3-methylindole] is a mild oxidant and 

brominating reagent that leads to polypeptide cleavage on the C-terminal side 

of tryptophan residues). 

<br>

<hr>

<h2><a name="Casp1"></a><a href="http://www.uniprot.org/uniprot/P29466">Caspase 1:</a></h2>

Caspase-1 is acting on Interleukin-1 beta [Precursor] (<a href="http://www.uniprot.org/uniprot/P01584">P01584</a>) to release it by specific cleavage at 116-Asp-|-Ala-117 (YVHDA) and 27-Asp-|-Gly-28 (EADG) bonds. It also hydrolyzes small-molecule substrate such as Ac-Tyr-Val-Ala-Asp-|-NHMec. Generally the substrate/enzyme interaction is located between the positions P4 and P1'. Various different patterns were proposed such as  YEVD|X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian et al., 1997</a>) or WEHD|X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Thornberry">Thornberry et al., 1997</a>), where X is any amino acid but Pro, Glu, Asp, Gln, Lys, Arg (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Stennicke">Stennicke et al., 2000</a>, <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian et al., 1997</a>). The pattern implemented for PeptideCutter 

considers an extended rule based on the study by <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Earnshaw">Earnshaw et al., 1999</a>, to optimise the caspase-1 endoproteolytic specificity, and can be found in the table at the end of this document,  describing the possible variations on the different interacting sites from P4 to P'1.

<br>

<hr>

<h2><a name="Casp2"></a><a href="http://www.uniprot.org/uniprot/P42575">Caspase 2:</a></h2>

The present version of the PeptideCutter program considers only the preferred 

peptide substrate sites as summarized in<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Earnshaw"> 

Earnshaw et al., (1999)</a>: Cleavage preferentially occurs at sites the positions 

P4 to P1'of which are composed of DVAD<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian 

et al.,1997</a>) or DEHD<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Thornberry">Thornberry 

et al.,1997</a>), where X is any amino acid but Pro, Glu, Asp, Gln, Lys, Arg (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Stennicke">Stennicke et al., 2000</a>, <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian et al., 1997</a>). <br>

<hr>

<h2><a name="Casp3"></a><a href="http://www.uniprot.org/uniprot/P42574">Caspase 3:</a></h2>

The present version of the PeptideCutter program considers only the preferred 

peptide substrate sites as summarized in<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Earnshaw"> 

Earnshaw et al., (1999)</a>: Cleavage preferentially occurs at sites the positions 

P4 to P1'of which are composed of DMQD<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian 

et al.,1997</a>) or DEVD<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Thornberry">Thornberry 

et al.,1997</a>), where X is any amino acid but Pro, Glu, Asp, Gln, Lys, Arg (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Stennicke">Stennicke et al., 2000</a>, <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian et al., 1997</a>). <br>

<hr>

<h2><a name="Casp4"></a><a href="http://www.uniprot.org/uniprot/P49662">Caspase 4:</a></h2>

The present version of the PeptideCutter program considers only the preferred 

peptide substrate sites as summarized in<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Earnshaw"> 

Earnshaw et al., (1999)</a>: Cleavage preferentially occurs at sites the positions 

P4 to P1'of which are composed of LEVD<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian 

et al.,1997</a>) or (W/L)EHD<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Thornberry">Thornberry 

et al.,1997</a>), where X is any amino acid but Pro, Glu, Asp, Gln, Lys, Arg (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Stennicke">Stennicke et al., 2000</a>, <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian et al., 1997</a>). <br>

<hr>

<h2><a name="Casp5"></a><a href="http://www.uniprot.org/uniprot/P51878">Caspase 5:</a></h2>

The present version of the PeptideCutter program considers only the preferred 

peptide substrate sites as summarized in<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Earnshaw"> 

Earnshaw et al., (1999)</a>: Cleavage preferentially ooccurs at sites the positions 

P4 to P1' of which are composed of (W/L)EHD<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Thornberry">Thornberry 

et al.,1997</a>). <br>

<hr>

<h2><a name="Casp6"></a><a href="http://www.uniprot.org/uniprot/P55212">Caspase 6:</a></h2>

The present version of the PeptideCutter program considers only the preferred 

peptide substrate sites as summarized in<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Earnshaw"> 

Earnshaw et al., (1999)</a>: Cleavage preferentially occurs at sites the positions 

P4 to P1' of which are composed of VEID<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian 

et al.,1997</a>) or VEHD<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Thornberry">Thornberry 

et al.,1997</a>), where X is any amino acid but Pro, Glu, Asp, Gln, Lys, Arg (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Stennicke">Stennicke et al., 2000</a>, <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian et al., 1997</a>). <br>

<hr>

<h2><a name="Casp7"></a><a href="http://www.uniprot.org/uniprot/P55210">Caspase 7:</a></h2>

The present version of the PeptideCutter program considers only the preferred 

peptide substrate sites as summarized in<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Earnshaw"> 

Earnshaw et al., (1999)</a>: Cleavage preferentially ooccurs at sites the positions 

P4 to P1' of which are composed of DEVD<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian 

et al.,1997</a>;<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Thornberry">Thornberry 

et al.,1997</a>), where X is any amino acid but Pro, Glu, Asp, Gln, Lys, Arg (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Stennicke">Stennicke et al., 2000</a>, <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian et al., 1997</a>). <br>

<hr>

<h2><a name="Casp8"></a><a href="http://www.uniprot.org/uniprot/Q14790">Caspase 8:</a></h2>

The present version of the PeptideCutter program considers only the preferred 

peptide substrate sites as summarized in<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Earnshaw"> 

Earnshaw et al., (1999)</a>: Cleavage preferentially occurs at sites the positions 

P4 to P1' of which are composed of IETD<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian 

et al.,1997</a>) or LETD<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Thornberry">Thornberry 

et al.,1997</a>), where X is any amino acid but Pro, Glu, Asp, Gln, Lys, Arg (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Stennicke">Stennicke et al., 2000</a>, <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian et al., 1997</a>). <br>

<hr>

<h2><a name="Casp9"></a><a href="http://www.uniprot.org/uniprot/P55211">Caspase 9:</a></h2>

The present version of the PeptideCutter program considers only the preferred 

peptide substrate sites as summarized in<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Earnshaw"> 

Earnshaw et al., (1999)</a>: Cleavage preferentially occurs at sites the positions 

P4 to P1' of which are composed of LEHD<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Thornberry">Thornberry 

et al.,1997</a>). <br>

<hr>



<h2><a name="Casp10"></a><a href="http://www.uniprot.org/uniprot/Q92851">Caspase 10:</a></h2>

The present version of the PeptideCutter program considers only the preferred 

peptide substrate sites as summarized in<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Earnshaw"> 

Earnshaw et al., (1999)</a>: Cleavage preferentially occurs at sites the positions 

P4 to P1'of which are composed of IEAD<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Talanian">Talanian 

et al.,1997</a>). <br>

<hr>



<a name="Ch_hi"></a>

<a name="Ch_lo"></a>

<h2><a name="Chym"></a><a href="http://enzyme.expasy.org/EC/3.4.21.1">Chymotrypsin:</a></h2>

Chymotrypsin preferentially cleaves at Trp, Tyr and Phe in position P1(high specificity) and to a lesser extent (taken into account when dealing with low specificity chymotrypsin) at Leu, Met and His in position P1 (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">Keil, 1992</a>). Exceptions to these rules are the following: When Trp is found in position P1, the cleavage is blocked when Met or Pro are found in position P1'at the same time. Furthermore, Pro in position P1' nearly fully blocks the cleavage independent of the amino acids found in position P1.  When Met is found in position P1, the cleavage is blocked by the presence of Tyr in position P1'. Finally, when His is located in position P1, the presence of Asp, Met or Trp also blocks the cleavage . 

<br><hr>



<h2><a name="Clost"></a><a href="http://enzyme.expasy.org/EC/3.4.22.8">Clostripain (Clostridiopeptidase B):</a></h2>

Clostripain cleaves preferentially at the carboxyl group of arginine residues, i.e. Arg in position P1 (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">Keil, 1992</a>). This cleavage is not strict, especially when the time of proteolysis is short or when using specific native protein substrates. The cleavage of lysyl-bonds has been reported rarely. Clostripain accepts well substrates containing Lys instead of Arg, however reaction rates are very low in comparison to reactions with Arg containing substrates.

This enzymes is reported to be sensitive to the composition of the potential substrate site. However,  no rules can be defined. Probably Glu and Asp in position P1' protect against cleavage, as well as the accumulation of positive charge in the positions P1' to P4'.

<br><hr>



<h2><a name="CNBr"></a>CNBr:</h2>

<p>CNBr cleaves at Met in position P1. (Information from the Cutter of the Prolysis 

  program/Universite de Tours). When CNBr is not applied in large excess, the 

  cleavage may become incomplete. <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Schroeder">Schroeder 

  et al. (1969)</a> discuss resistance to CNBr that results from Ser or Thr being 

  located in P1' (In this case, the Met residue is converted into a 

  homoserine-residue, thus preventing a cleavage) or P2. In general, this 

  type of cleavage-blocking is prevented by using CNBr in large excess in relation 

  to the number of Met- residues in the sequence. <br>

</p>

<hr>



<h2><a name="Enter"></a><a href="http://enzyme.expasy.org/EC/3.4.21.9">Enterokinase:</a></h2>

<p>

Enterokinase is a serine protease that recognizes the amino acid sequence -Asp-Asp-Asp-Asp-Lys-<b>|</b>-X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Roche">Roche</a>)

 with a high specificity. The enterokinase activates its natural substrate trypsinogen and releases trypsin

by cleavage at the C-terminal end of this sequence. The aspartic acid residues can be substituted by glutamic acid.

Note that PeptideCutter does not take into account position P5. The implemented motif

for Enterokinase is therefore [DE][DE][DE][DE]K-X.



  <br>

</p>

<hr>



<h2><a name="Xa"></a><a href="http://enzyme.expasy.org/EC/3.4.21.6">Factor Xa:</a></h2>

(Coagulation factor Xa) Factor Xa is prepared by the activation of its precursor, 

the mostly inactive Factor X, by the hydrolysis of a specific peptide bond in 

the amino-terminal region of the heavy chain (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Fujikawa">Fujikawa 

et al., 1975</a>). Highly specific for cleavage at Arg in position P1 and Gly 

in position P2. In general position P3 is occupied by a negatively charged residue 

(Glu or Asp) and position P4 may be hydrophobic (Ile or Ala). The composition 

of the P'-sites do not seem to considerably influence the cleavage. <br>

<hr>



<h2><a name="HCOOH"></a>Formic acid:</h2>

Cleaves at Asp in position P1 (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Li">Li et al., 2001</a>). 

<br>

<hr>





<h2><a name="Glu"></a> <a href="http://enzyme.expasy.org/EC/3.4.21.19">Glutamyl endopeptidase:</a></h2>

Mainly three different types are commercially available (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Birktoft">Birktoft 

and Breddam, 1994</a>): GluBl (from <i>Bacillus licheniformis</i>, GluSGB (<i>Streptomyces 

griseus</i>), GluV8 (<i>Streptococcus aureus</i>, strain<i> V8</i>). All of them 

preferentially cleave at Glu in position P1. When Glu and Asp are found in a directly 

neighbouring positions the cleavage at Glu in position P1 is preferred a 100-(GluSGP) 

to around 1000-fold (other two) in relation to Asp in position P1. The nature 

of the reaction buffer (whether bicarbonate or phosphate) does not seem to have 

an influence on this ratio, however it has been shown that reactivity in general 

is enhanced with phosphate in comparison to bicarbonate (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Houmard">Houmard 

and Drapeau,1972</a>). Generally preferred composition of cleavage site: Asp in 

position P4, Ala/Val in position P3, Pro/Val in position P2 (GluSGP) or Phe in 

position P2 (GluBL/GluV8). Disfavouring of Pro in position P3, position P1' and 

position P2' as well as Asp at position P1'. <br>

<hr>





<h2><a name="GranB"></a><a href="http://enzyme.expasy.org/EC/3.4.21.79">Granzyme B:</a></h2>

The present version of the PeptideCutter program considers only the preferred 

peptide substrate sites as summarized in<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Earnshaw"> 

Earnshaw et al., (1999)</a>: Cleavage preferentially occurs at sites the positions 

P4 to P1'of which are composed of IEPD<b>|</b>X (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Thornberry">Thornberry 

et al.,1997</a>). <br>

<hr>



<h2><a name="Hydro"></a>Hydroxylamine (NH2OH):</h2>

Cleaves at Asn in position P1 and Gly in position P1' (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Bornstein">Bornstein &amp; Balian</a>)

<hr>

<h2><a name="Iodo"></a>Iodosobenzoic acid:</h2>

Cleaves at Trp in position P1. (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Han">Han et al., 1983</a>)

<hr>

<h2><a name="LysC"></a><a href="http://enzyme.expasy.org/EC/3.4.21.50">LysC 

   Lysyl endopeptidase (Achromobacter proteinase I):</a></h2>

Cleaves at Lys in position P1 <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">(Keil, 

1992)</a>. 

<br>

<hr>

<h2><a name="LysN"></a><a href="http://enzyme.expasy.org/EC/3.4.24.20">LysN

   Peptidyl-Lys metalloendopeptidase</a></h2>

Cleaves at Lys in position P1' <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">(Keil,

1992)</a>.

<br>

<hr>



<h2><a name="Elast"></a><a href="http://www.uniprot.org/uniprot/P08246">Neutrophil elastase

</a></h2>

Cleaves at Val or Ala in position P1 (<a href="http://enzyme.expasy.org/EC/3.4.21.37">EC 3.4.21.37</a>).

<br>

<hr>



<h2><a name="NTCB"></a>NTCB +Ni (2-nitro-5-thiocyanobenzoic acid ):</h2>

Cleaves at Cys in position P1' (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Degani">Degani and Patchornik, 1974</a>).

<br><hr>



<a name="Pn1.3"></a>

<a name="Pn2"></a>

<h2><a name="Peps"></a><a href="http://enzyme.expasy.org/EC/3.4.23.1">Pepsin:</a></h2>

Pepsin preferentially cleaves at Phe, Tyr, Trp and Leu in position P1 or P1'(<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">Keil, 

1992</a>). Negative effects on cleavage are excerted by Arg, Lys and His in position 

P3 and Arg in position P1. Pro has favourable effects when being located in position 

P4 and position P3, but unfavourable ones when found in positions P2 to P3'. Cleavage 

is more specific at pH 1.3. Then pepsin preferentially cleaves at Phe and Leu 

in position P1 with negligible cleavage for all other amino acids in this position. 

This specificity is lost at pH &gt;= 2. <br>

<hr>



<h2><a name="Pro"></a><a href="http://enzyme.expasy.org/EC/3.4.21.26">Proline-endopeptidase:</a></h2>

Proline-endopeptidase preferentially cleaves at Pro in position P1 (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">Keil, 1992</a>). Proline-endopeptidase may also accept Ala in position P1. With Pro in position P1 the activity is blocked when another Pro is at position P1'. In most cases a basic amino acid (Lys, His, Arg) is found in position P2. It was suggested that this feature is obligatory. Some discrepancies concerning the cleavage specificity were observed in individual cases, however this may be due to impurities or to the fact that Pro-endopeptidases came from different sources and may not be identical.  

<br><em>NOTE: Proline-endopeptidase was reported to cleave only substrates

whose sequences do not exceed 30 amino acids. An unusual beta-propeller domain regulates proteolysis: see <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Fulop">Fulop et al., 1998</a>.</em>

<hr>



<h2><a name="ProtK"></a><a href="http://enzyme.expasy.org/EC/3.4.21.64">Proteinase K:</a></h2>

Proteinase K preferentially cleaves at aliphatic of aromatic amino acid residues in position P1 <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">(Keil, 1992)</a>. Ala in position P2 enhances the cleavage . The specificity of proteinase K is not always unambiguous.

<br><hr>



<h2><a name="Staph"></a><a href="http://enzyme.expasy.org/EC/3.4.21.19">Staphylococcal peptidase I:</a></h2>

Preferentially cleaves at Glu in position P1, but also, although to a lesser extent, 

at Asp in position P1 (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">Keil, 

1992</a>). In rare cases Ser can be accepted in position P1. In general, specificity 

depends strongly on the experimental conditions: the exchange of buffer (not necessarily 

of pH) can change the cleavage behaviour. When two Glu residues are found in directly 

neighbouring positions, Staphylococcal peptidase I prefers to cut at Glu in position 

P1 with another Glu in position P1' instead of the second Glu being located in 

position P2. <br>

<hr>

<h2><a name="TEV"></a><a href="http://enzyme.expasy.org/EC/3.4.22.44">Tobacco etch virus protease:</a></h2>

TEV protease is the common name for the 27 kDa catalytic domain of the Nuclear

Inclusion a (NIa) protein encoded by the tobacco etch virus (TEV). Because its

sequence specificity is far more stringent than that of factor Xa, thrombin, or

enterokinase, TEV protease is a very useful reagent for cleaving fusion proteins. It is

also relatively easy to overproduce and purify large quantities of the enzyme.

TEV protease recognizes a linear epitope of the general form E-Xaa-Xaa-Y-Xaa-Q-(G/S),

with cleavage occurring between Q and G or Q and S. The most commonly used

sequence is ENLYFQG (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Waugh2002">Waugh, 

2002</a>, <a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#WaughFAQ">Waugh, TEV protease FAQ)</a>).

Note that PeptideCutter does not take into account positions P5 and P6. The implemented motif

for TEV protease is therefore XYXQ-[GS].



<br>

<hr>



<h2><a name="Therm"></a><a href="http://enzyme.expasy.org/EC/3.4.24.27">Thermolysin:</a></h2>

Thermolysin preferentially cleaves sites with bulky and aromatic residues (Ile, 

Leu, Val, Ala, Met, Phe) in position P1' (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">Keil, 

1992</a>). Cleavage is favoured with aromatic sites in position P1 but hindered 

with acidic residues in position P1. Pro blocks when located inposition P2' but 

not when found in position P1. <br>

<hr>



<h2><a name="Throm"></a><a href="http://enzyme.expasy.org/EC/3.4.21.5">Thrombin:</a></h2>

Preferentially cleaves at Arg in position P1 (<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">Keil, 

1992</a>).The natural substrate of thrombin is fibrinogen. Optimum cleavage sites 

are when Arg in position P1 and Gly in position P2 and position P1'. Likewise, 

when hydrophobic residues are found in position P4 and position P3 , Pro in position 

P2, Arg in position P1, and non-acidic amino-acids in position P1' and position 

P2'. A very important residue for its natrual substrate fibrinogen is an Asp in 

P10 (but this site is neglected in the PeptideCutter program). <br>

<hr>



<h2><a name="Tryps"></a><a href="http://enzyme.expasy.org/EC/3.4.21.4">Trypsin:</a></h2>

Preferentially cleaves at Arg and Lys in position P1 with higher rates for Arg 

(<a href="http://web.expasy.org/peptide_cutter/peptidecutter_references.html#Keil1992">Keil, 1992</a>), especially 

at high pH (but treated equally in the program). Pro usually blocks the action 

when found in position P1', but not when Lys is in position P1 and Trp is in position 

P2 at the same time. This blocking of cleavage exerted by Pro in position P1' 

is also negligible when Arg is in position P1 and Met is in position P2 at the 

same time (other reports say that the block exhibited by Pro can be circumvented 

by Glu being in P2). <br>

<br>

Furthermore, if Lys is found in position P1 the following situation considerably 

block the action of trypsin: Either Asp in position P2 and Asp in position P1' 

or Cys in position P2 and Asp in position P1'or Cys in position P2 and His in 

position P1' or Cys in position P2 and Tyr in position P1'. A likewise considerable 

block of trypsin action is seen , when Arg is in P1 and the following situations 

are found: Either Arg in position P2 and His in position P1' or Cys in position 

P2 and Lys in position P1' or Arg in position P2 and Arg in positionP1'.<br>

<br>



This Arg/Lys specificity is seen very nicely with pure alpha- and beta trypsins. Trypsin preparations with traces of "pseudotrypsin" also cleave considerably at the following amino acids in P1:Phe (except with Glu or Pro in P1'), Tyr (except with Pro and Arg in P1') and Trp (except with Ile, Lys, Pro Val and Trp in P1') Met (with Ala, His, Met, Gln, Ser, Val and Trp in P1') and Cys (with Phe, Gly, Ile, Leu, Val and Trp in P1').

<br><hr>





<hr>

<h2> Summary of the cleavage rules:</h2>

<br>

<hr>



<h2>Cleavage rules</h2>

The following enzymes potentially cleave when the respective compositions of the cleavage sites are found. However, there also are some <a href="http://web.expasy.org/peptide_cutter/peptidecutter_enzymes.html#exceptions"> exceptions</a>.

<!--<table ><tr><td>-->

<table frame="void" border="" rules="rows" cellpadding="5">

<tbody><tr><td><b>Enzyme name</b></td><td><b>P4</b></td><td><b>P3</b></td><td><b>P2</b></td><td><b>P1</b></td><td><b>P1'</b></td><td><b>P2'</b></td></tr>

<tr><td>Arg-C proteinase</td><td>-</td><td>-</td><td>-</td><td>R</td><td>-</td><td>-</td></tr>

<tr><td>Asp-N endopeptidase</td><td>-</td><td>-</td><td>-</td><td>-</td><td>D</td><td>-</td></tr>

<tr><td>BNPS-Skatole</td><td>-</td><td>-</td><td>-</td><td>W</td><td>-</td><td>-</td></tr>

<tr><td>Caspase 1</td>

          <td>F, W, Y, or L</td>

          <td>-</td><td>H, A or T</td><td>D</td><td>not P, E, D, Q, K or  R </td><td>-</td></tr>

<tr><td>Caspase 2</td><td>D</td><td>V</td><td>A</td><td>D</td><td>not P, E, D, Q, K or R</td><td>-</td></tr>

<tr><td>Caspase 3</td><td>D</td><td>M</td><td>Q</td><td>D</td><td>not P, E, D, Q, K or R</td><td>-</td></tr>

<tr><td>Caspase 4</td><td>L</td><td>E</td><td>V</td><td>D</td><td>not P, E, D, Q, K or R</td><td>-</td></tr>

<tr><td>Caspase 5</td>

          <td>L or W</td>

          <td>E</td><td>H</td><td>D</td><td>-</td><td>-</td></tr>

<tr><td>Caspase 6</td><td>V</td><td>E</td>

          <td>H or I</td>

          <td>D</td><td>not P, E, D, Q, K or  R</td><td>-</td></tr>

<tr><td>Caspase 7</td><td>D</td><td>E</td><td>V</td><td>D</td><td>not P, E, D, Q, K or R</td><td>-</td></tr>

<tr><td>Caspase 8</td>

          <td>I or L</td>

          <td>E</td><td>T</td><td>D</td><td>not P, E, D, Q, K or R</td><td>-</td></tr>

<tr><td>Caspase 9</td><td>L</td><td>E</td><td>H</td><td>D</td><td>-</td><td>-</td></tr>

<tr><td>Caspase 10</td><td>I</td><td>E</td><td>A</td><td>D</td><td>-</td><td>-</td></tr>

<tr><td rowspan="2">Chymotrypsin-high specificity (C-term to [FYW], not before P)</td><td>-</td><td>-</td><td>-</td>

          <td>F or Y</td>

          <td>not P</td><td>-</td></tr>

<tr><td>-</td><td>-</td><td>-</td><td>W</td>

          <td>not M or P</td>

          <td>-</td></tr>

<tr><td rowspan="4">Chymotrypsin-low specificity (C-term to [FYWML], not before P)</td><td>-</td><td>-</td><td>-</td>

          <td>F,L or Y</td>

          <td>not P</td><td>-</td></tr>

<tr><td>-</td><td>-</td><td>-</td><td>W</td>

          <td>not M or P</td>

          <td>-</td></tr>

<tr><td>-</td><td>-</td><td>-</td><td>M</td>

          <td>not P or Y</td>

          <td>-</td></tr>

<tr><td>-</td><td>-</td><td>-</td><td>H</td>

          <td>not D,M,P or W</td>

          <td>-</td></tr>

<tr><td>Clostripain (Clostridiopeptidase B)</td><td>-</td><td>-</td><td>-</td><td>R</td><td>-</td><td>-</td></tr>

<tr><td>CNBr</td><td>-</td><td>-</td><td>-</td><td>M</td><td>-</td><td>-</td></tr>

<tr><td>Enterokinase</td><td>D or E</td><td>D or E</td><td>D or E</td><td>K</td><td>-</td><td>-</td></tr>

<tr><td>Factor Xa</td>

          <td>A,F,G,I,L,T,V or M</td>

          <td>D or E</td>

          <td>G</td><td>R</td><td>-</td><td>-</td></tr>

<tr><td>Formic acid</td><td>-</td><td>-</td><td>-</td><td>D</td><td>-</td><td>-</td></tr>

<tr><td>Glutamyl endopeptidase</td><td>-</td><td>-</td><td>-</td><td>E</td><td>-</td><td>-</td></tr>

<tr><td>GranzymeB</td><td>I</td><td>E</td><td>P</td><td>D</td><td>-</td><td>-</td></tr>

<tr><td>Hydroxylamine</td><td>-</td><td>-</td><td>-</td><td>N</td><td>G</td><td>-</td></tr>

<tr><td>Iodosobenzoic acid</td><td>-</td><td>-</td><td>-</td><td>W</td><td>-</td><td>-</td></tr>

<!--</table></tr>-->

<tr><td>LysC</td><td>-</td><td>-</td><td>-</td><td>K</td><td>-</td><td>-</td></tr>

<tr><td>Neutrophil elastase</td><td>-</td><td>-</td><td>-</td><td>A or V</td><td>-</td><td>-</td></tr>

<tr><td>NTCB (2-nitro-5-thiocyanobenzoic acid)</td><td>-</td><td>-</td><td>-</td><td>-</td><td>C</td><td>-</td></tr>

<tr><td rowspan="2">Pepsin (pH1.3)</td><td>-</td><td>not H,K, or R</td><td>not P</td><td>not R</td><td>F,L,W or Y</td><td>not P</td></tr>

<tr><td>-</td><td>not H,K, or R</td><td>not P</td><td>F,L,W or Y</td><td>-</td><td>not P</td></tr>



<tr><td rowspan="2">Pepsin (pH&gt;2)</td><td>-</td><td>not H,K or R</td><td>not P</td><td>not R</td><td>F or L</td><td>not P</td></tr>

<tr><td>-</td><td>not H,K or R</td><td>not P</td><td>F or L</td><td>-</td><td>not P</td></tr>



<tr><td>Proline-endopeptidase</td><td>-</td><td>-</td>

    <td>H,K or R</td>

    <td>P</td><td>not P</td><td>-</td></tr>

<tr><td>Proteinase K</td><td>-</td><td>-</td><td>-</td>

    <td>A,E,F,I,L,T,V,W or Y</td>

    <td>-</td><td>-</td></tr>

<tr><td>Staphylococcal peptidase I</td><td>-</td><td>-</td><td>not E</td><td>E</td><td>-</td><td>-</td></tr>

<tr><td>Thermolysin</td><td>-</td><td>-</td><td>-</td>

    <td>not D or E</td>

    <td>A,F,I,L,M or V</td>

    <td>-</td></tr>

<tr><td rowspan="2">Thrombin</td><td>-</td><td>-</td><td>G</td><td>R</td><td>G</td><td>-</td></tr>

    <tr><td>A,F,G,I,L,T,V or M</td>

    <td>A,F,G,I,L,T,V,W or A</td>

    <td>P</td><td>R</td>

    <td>not D or E</td>

    <td>not DE</td></tr>

<tr><td rowspan="3">Trypsin (please note the<a href="http://web.expasy.org/peptide_cutter/peptidecutter_enzymes.html#exceptions"> exceptions</a>!)</td><td>-</td><td>-</td><td>-</td>

    <td>K or R</td>

    <td>not P</td><td>-</td></tr>

<tr><td>-</td><td>-</td><td>W</td><td>K</td><td>P</td><td>-</td></tr>

<tr><td>-</td><td>-</td><td>M</td><td>R</td><td>P</td><td>-</td></tr>

<tr><td><b>Enzyme name</b></td><td><b>P4</b></td><td><b>P3</b></td><td><b>P2</b></td><td><b>P1</b></td><td><b>P1'</b></td><td><b>P2'</b></td></tr>

</tbody></table>







<br><br><h2><a name="exceptions">The exception rules:</a></h2><br>

The above cleavage rules do not  apply, i.e. no cleavage occurs, with the following compositions of the cleavage sites:



<!--<table ><tr><td>-->

<table frame="void" border="" rules="rows" cellpadding="5">

<tbody><tr><td><b>Enzyme name</b></td><td><b>P4</b></td><td><b>P3</b></td><td><b>P2</b></td><td><b>P1</b></td><td><b>P1'</b></td><td><b>P2'</b></td></tr>

<tr><td rowspan="4"><a name="Tryps_exceptions">Trypsin</a></td><td>-</td><td>-</td>

            <td>C or D</td>

            <td>K</td><td>D</td><td>-</td></tr>

<tr><td>-</td><td>-</td><td>C</td><td>K</td><td>H or Y</td><td>-</td></tr>

<tr><td>-</td><td>-</td><td>C</td><td>R</td><td>K</td><td>-</td></tr>

<tr><td>-</td><td>-</td><td>R</td><td>R</td><td>H or R</td><td>-</td></tr>



</tbody></table>

<!-- sib_body -->
<br clear="left">
</div>
<!-- ********************************************************************************************** -->
<!-- START OF FOOTER information at the bottom of the page -->
<div id="sib_footer">			
  <!-- This part of the footer is _left_ justified and should contain SIB on the very left -->	
  <div id="sib_footer_content"> 
    <a href="http://www.isb-sib.ch/">SIB Swiss Institute of Bioinformatics</a>&nbsp;|&nbsp;<a href="http://www.expasy.org/disclaimer.html">Disclaimer</a>

	<!-- This part of the footer is _right_ justified -->
	<div id="sib_footer_right">	
      <a href="http://web.expasy.org/peptide_cutter/peptidecutter_enzymes.html#TOP" id="sib_footer_gototop"><span style="padding-left: 10px">Back to the Top</span></a>				
    </div>
  </div>
</div>
<!-- END OF FOOTER -->
</div> <!-- sib_container -->
<script type="text/javascript" async="" src="./expasy_peptidecuttr_enzymes_files/ga.js"></script><script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-23625305-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script><!-- automatically added by Apache -->
<!-- automatically added by Apache -->
 











</body></html>